bpo-36974: expand call protocol documentation

jdemeyer opened the pull request:
CC @encukou 

I'm also adding Petr Viktorin as contributor for vectorcall in the "what's new" section.

<!-- issue-number: [bpo-36974](https://bugs.python.org/issue36974) -->
https://bugs.python.org/issue36974
<!-- /issue-number -->


Automerge-Triggered-By: @encukou

Automerge-Triggered-By: @encukou

jdemeyer committed:
bpo-36974: separate docs for call protocol

jdemeyer committed:
bpo-36974: expand call protocol documentation

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+This convention is not only used by *tp_call*:
+also :c:member:`~PyTypeObject.tp_new` and :c:member:`~PyTypeObject.tp_init`
+pass arguments this way
```
```suggestion
also pass arguments this way
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+but *kwargs* may be *NULL* if there are no keyword arguments.
+
+This convention is not only used by *tp_call*:
+also :c:member:`~PyTypeObject.tp_new` and :c:member:`~PyTypeObject.tp_init`
```
```suggestion
:c:member:`~PyTypeObject.tp_new` and :c:member:`~PyTypeObject.tp_init`
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+A call is made using a tuple for the positional arguments
+and a dict for the keyword arguments, just like ``f(*args, **kwargs)``
+in the Python language.
+*args* must be non-NULL
```
Why are these lines so short?

jdemeyer replied:
It doesn't matter in the built documentation. I personally prefer breaking such lines logically, instead of at an arbitrary number of characters. So I tend to use a newline at the end of a sentence. If this is considered bad style, I can fix it.

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+     To get the actual number of positional arguments from *nargsf*,
+     use :c:func:`PyVectorcall_NARGS`.
+   - *kwnames* is a tuple containing the names of the keyword arguments,
+     in other words the keys of the kwargs dict.
```
```suggestion
     in other words, the keys of the kwargs dict.
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+     :const:`PY_VECTORCALL_ARGUMENTS_OFFSET` flag.
+     To get the actual number of positional arguments from *nargsf*,
+     use :c:func:`PyVectorcall_NARGS`.
+   - *kwnames* is a tuple containing the names of the keyword arguments,
```
```suggestion
   - *kwnames* is a tuple containing the names of the keyword arguments;
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+.. warning::
+
+   As rule of thumb, CPython will internally use the vectorcall
+   protocol if the callable supports it. However, this is not a hard rule,
```
```suggestion
   protocol if the callable supports it. However, this is not a hard rule;
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+   As rule of thumb, CPython will internally use the vectorcall
+   protocol if the callable supports it. However, this is not a hard rule,
+   *tp_call* may be used instead.
+   Therefore, a class supporting vectorcall must also set
```
Maybe "implement" instead of "set"?
```suggestion
   Therefore, a class supporting vectorcall must also implement
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+   Call a callable Python object *callable*, with arguments given by the
+   tuple *args*, and named arguments given by the dictionary *kwargs*.
+
+   *args* must not be *NULL*, use an empty tuple if no arguments are needed.
```
```suggestion
   *args* must not be *NULL*; use an empty tuple if no arguments are needed.
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+   Call the method named *name* of object *obj* with a variable number of C
+   arguments.  The C arguments are described by a :c:func:`Py_BuildValue` format
+   string that should  produce a tuple.
```
```suggestion
   string that should produce a tuple.
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+.. c:function:: PyObject* PyObject_CallFunctionObjArgs(PyObject *callable, ..., NULL)
+
+   Call a callable Python object *callable*, with a variable number of
+   :c:type:`PyObject\*` arguments.  The arguments are provided as a variable number
```
```suggestion
   :c:type:`PyObject \*` arguments.  The arguments are provided as a variable number
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+.. c:function:: PyObject* PyObject_CallMethodObjArgs(PyObject *obj, PyObject *name, ..., NULL)
+
+   Calls a method of the Python object *obj*, where the name of the method is given as a
```
```suggestion
   Call a method of the Python object *obj*, where the name of the method is given as a
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+   Calls a method of the Python object *obj*, where the name of the method is given as a
+   Python string object in *name*.  It is called with a variable number of
+   :c:type:`PyObject\*` arguments.  The arguments are provided as a variable number
```
```suggestion
   :c:type:`PyObject \*` arguments.  The arguments are provided as a variable number
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+.. c:function:: PyObject* _PyObject_FastCallDict(PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwdict)
+
+   Call *callable* with positional arguments passed exactly as in the vectorcall_ protocol
```
```suggestion
   Call *callable* with positional arguments passed exactly as in the vectorcall protocol,
```

jdemeyer replied:
What's wrong with the reference `vectorcall_`?

epicfaace replied:
I don't see `vectorcall_` anywhere else in the code or doc -- is the underscore some kind of rst syntax? I guess this is moot anyway since this part was removed.

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+   If *op* does not support the vectorcall protocol (either because the type
+   does not or because the specific instance does not), return *NULL*.
+   Otherwise, return the vectorcall function pointer stored in *op*.
+   This function never sets an exception.
```
```suggestion
   This function never raises an exception.
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+.. c:function:: PyObject* _PyObject_VectorcallMethod(PyObject *name, PyObject *const *args, size_t nargsf, PyObject *kwnames)
+
+   Call a method using the vectorcall calling convention. The name of the method
+   is given as Python string *name*. The object whose method is called is
```
```suggestion
   is given as the Python string *name*. The object whose method is called is
```

epicfaace commented on Doc/c-api/call.rst:
```diff
@@ -0,0 +1,344 @@
+
+   Call a method using the vectorcall calling convention. The name of the method
+   is given as Python string *name*. The object whose method is called is
+   *args[0]* and the *args* array starting at *args[1]* represents the arguments
```
```suggestion
   *args[0]*, and the *args* array starting at *args[1]* represents the arguments
```

epicfaace commented on Doc/c-api/exceptions.rst:
```diff
@@ -695,13 +695,17 @@ The following functions are used to create and modify Unicode exceptions from C.
 level, both in the core and in extension modules.  They are needed if the
 recursive code does not necessarily invoke Python code (which tracks its
 recursion depth automatically).
+It's also not needed for *tp_call* implementations
```
```suggestion
They're also not needed for *tp_call* implementations
```

jdemeyer committed:
bpo-36974: review by Ashwin Ramaswami

encukou committed:
Reorganization and rewording

- Repeat the tp_call signature, for clarity
- Warn that vectorcall API is provisional
- Make the existing warning smaller, but move the explanation
  up a bit
- Group calling API together and provide a summary table
- Put smaller vectorcall-related sections under vectorcall
- Minor rewordings

encukou committed:
Address review comments

- Adjust vectorcall warning to not talk about a single function
- Add underscore to _PyObject_CallOneArg
- Include PyObject_CallFunction and _PyObject_CallMethodNoArgs
- Summarize _PyObject_FastCallDict's args handling as "vectorcall"
- Use "name" rather chan "char*" when the name is a Python object

encukou committed:
Merge branch 'master' into doc_call

encukou committed:
Merge branch origin/master into doc_call
